x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportGetDeploymentStatsAction.java

```diff
@@ -0,0 +1,135 @@
+                (l, r) -> {
+                    l.getNodeStats().addAll(r.getNodeStats());
+                    return l;
+                }));
+
+        List<GetDeploymentStatsAction.Response.AllocationStats> bunchedAndSorted =
+            mergedNodeStatsByModel.values().stream()
+                .sorted(Comparator.comparing(GetDeploymentStatsAction.Response.AllocationStats::getModelId))
+                .collect(Collectors.toList());
+
+        return new GetDeploymentStatsAction.Response(taskOperationFailures,
+            failedNodeExceptions,
+            bunchedAndSorted,
+            bunchedAndSorted.size());
+    }
+
+    @Override
+    protected void doExecute(Task task, GetDeploymentStatsAction.Request request,
+                             ActionListener<GetDeploymentStatsAction.Response> listener) {
+
+        String[] tokenizedRequestIds = Strings.tokenizeToStringArray(request.getDeploymentId(), ",");
+        ExpandedIdsMatcher.SimpleIdsMatcher idsMatcher =
+            new ExpandedIdsMatcher.SimpleIdsMatcher(tokenizedRequestIds);
+
+        TrainedModelAllocationMetadata allocation = TrainedModelAllocationMetadata.fromState(clusterService.state());
+        List<String> matchedDeploymentIds = new ArrayList<>();
+        Set<String> taskNodes = new HashSet<>();
+
+        for (var routingEntry : allocation.modelAllocations().entrySet()) {
+            if (idsMatcher.idMatches(routingEntry.getKey())) {
+                matchedDeploymentIds.add(routingEntry.getKey());
+
+                taskNodes.addAll(Arrays.asList(routingEntry.getValue().getStartedNodes()));  // TODO: limited to started nodes
+                logger.warn("Routing table " + Strings.toString(routingEntry.getValue()));
+                logger.warn("getting stats on nodes: " + taskNodes);
+            }
+        }
+
+        // check request has been satisfied
+        ExpandedIdsMatcher requiredIdsMatcher = new ExpandedIdsMatcher(tokenizedRequestIds, request.isAllowNoMatch());
+        requiredIdsMatcher.filterMatchedIds(matchedDeploymentIds);
+        if (requiredIdsMatcher.hasUnmatchedIds()) {
+            listener.onFailure(ExceptionsHelper.missingDeployment(requiredIdsMatcher.unmatchedIdsString()));
+            return;
+        }
+        if (matchedDeploymentIds.isEmpty()) {
+            listener.onResponse(new GetDeploymentStatsAction.Response(
+                Collections.emptyList(), Collections.emptyList(), Collections.emptyList(), 0L));
+            return;
+        }
+
+        request.setNodes(taskNodes.toArray(String[]::new));
+        request.setExpandedIds(matchedDeploymentIds);
+
+        super.doExecute(task, request, listener);
+    }
+
+    @Override
+    protected void taskOperation(GetDeploymentStatsAction.Request request, TrainedModelDeploymentTask task,
+                                 ActionListener<GetDeploymentStatsAction.Response.AllocationStats> listener) {
+        ModelStats stats = deploymentManager.getStats(task);
```

Comment:
This may throw due to `RoutingState` `STARTING`, `FAILED`, or `STOPPED`.

Obviously, for `STARTING` and `STOPPED`, I don't think this should fail, we should instead return a state. `FAILED` should probably not "fail" this either but indicate its state in the `NodeStats` object.

Reply:
The task operation is now only routed to nodes with a `STARTED` task. 

For the other `RoutingState`s such as `FAILED` those are merged with the responses from the tasks so that there is a response for all nodes even if the deployment failed on a node. In the case of a conflict where there there is a non-started routing state for a model and node pair and there a task response for the same model node pair then we take the allocation state when the request was initiated as the source of truth and choose that over the task response.